dockerunit-core/src/main/java/com/github/qzagarese/dockerunit/DockerUnitRule.java

```diff
@@ -0,0 +1,135 @@
+        List<DiscoveryProviderFactory> implementations = new ArrayList<>();
+        loader.forEach(impl -> {
+            logger.info("Found discovery provider factory of type " + impl.getClass().getSimpleName());
+            implementations.add(impl);
+        });
+        if (implementations.size() > 0) {
+            logger.info("Using discovery provider factory " + implementations.get(0).getClass().getSimpleName());
+            discoveryProvider = implementations.get(0).getProvider();
+        } else {
+            throw new RuntimeException("No discovery provider factory found. Aborting test.");
+        }
+        this.serviceContextName = serviceContextName;
+    }
+
+    @Override
+    public Statement apply(final Statement base, final Description description) {
+        return new Statement() {
+
+            @Override
+            public void evaluate() throws Throwable {
+                
+                logger.log(Level.INFO, "Performing service discovery for the following context: " + description.getDisplayName());
+                try {
+                    doSetup(description);
+                    base.evaluate(); // This will run the test.
+                } finally {
+                    logger.log(Level.INFO, "Cleaning up active services for the following context: " + description.getDisplayName());
+                    doTeardown();
+                }
+            }
+        };
+    }
+
+    private void doSetup(final Description description) {
+        UsageDescriptor descriptor = descriptorBuilder.buildDescriptor(description.getTestClass());
+        UsageDescriptor discoveryProviderDescriptor = descriptorBuilder.buildDescriptor(discoveryProvider.getDiscoveryConfig());
+      
+        // Build discovery context
+        ServiceContext discoveryContext = contextBuilder.buildContext(discoveryProviderDescriptor);
+        DockerUnitRule.this.discoveryContext = discoveryContext;
+        if(!discoveryContext.allHealthy()) {
+            throw new RuntimeException(discoveryContext.getFormattedErrors());
+        }
+
+        
+        // Create containers and perform discovery one service at the time
+        List<ServiceContext> serviceContexts = new ArrayList<>();
+        descriptor.getDependencies().forEach(usage -> {
+            ServiceContext context = contextBuilder.buildServiceContext(usage);
+            if(!context.allHealthy()) {
+                throw new RuntimeException(context.getFormattedErrors());
+            }
+            context = discoveryProvider.populateRegistry(context);
+            serviceContexts.add(context);
+        });
+        ServiceContext completeContext = mergeContexts(serviceContexts);
+        activeContexts.put(DockerUnitRule.this.serviceContextName, completeContext);
+        if(!completeContext.allHealthy()) {
+            throw new RuntimeException(completeContext.getFormattedErrors());
+        }
```

Comment:
Like above, feels to me that this could become a "cleaner" stream chain:
```
descriptor.getDependencies()
                .stream()
                .map(contextBuilder::buildServiceContext)
                .map(ctx -> {
                    if (!ctx.allHealthy()) {
                        throw new RuntimeException(ctx.getFormattedErrors());
                    }

                    return discoveryProvider.populateRegistry(ctx);
                })
                .collect(Collectors.toList());
```

Reply:
:+1: